include/particles/twophase_particle.tcc

```diff
@@ -0,0 +1,726 @@
+    status = false;
+  }
+  return status;
+}
+
+//! Assign particle permeability
+template <unsigned Tdim>
+bool mpm::TwoPhaseParticle<Tdim>::assign_permeability() {
+  bool status = true;
+  try {
+    // Check if material ptr is valid
+    if (this->material() != nullptr) {
+      // Porosity parameter
+      const double k_p =
+          std::pow(this->porosity_, 3) / std::pow((1. - this->porosity_), 2);
+      switch (Tdim) {
+        case (1): {
+          // Check if the permeability is valid
+          if (this->material()->template property<double>("k_x") < 0)
+            throw std::runtime_error("Material's permeability is invalid");
+          // Assign permeability
+          permeability_(0) =
+              this->material()->template property<double>("k_x") / k_p;
+          break;
+        }
+        case (2): {
+          // Check if the permeability is valid
+          if (this->material()->template property<double>("k_x") < 0 ||
+              this->material()->template property<double>("k_y") < 0)
+            throw std::runtime_error("Material's permeability is invalid");
+          // Assign permeability
+          permeability_(0) =
+              this->material()->template property<double>("k_x") / k_p;
+          permeability_(1) =
+              this->material()->template property<double>("k_y") / k_p;
+          break;
+        }
+        default: {
+          // Check if the permeability is valid
+          if (this->material()->template property<double>("k_x") < 0 ||
+              this->material()->template property<double>("k_y") < 0 ||
+              this->material()->template property<double>("k_z") < 0)
+            throw std::runtime_error("Material's permeability is invalid");
+          // Assign permeability
+          permeability_(0) =
+              this->material()->template property<double>("k_x") / k_p;
+          permeability_(1) =
+              this->material()->template property<double>("k_y") / k_p;
+          permeability_(2) =
+              this->material()->template property<double>("k_z") / k_p;
+          break;
+        }
+      }
+    } else {
+      throw std::runtime_error("Material is invalid");
+    }
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Map drag force
+template <unsigned Tdim>
+bool mpm::TwoPhaseParticle<Tdim>::map_drag_force_coefficient() {
+  bool status = true;
+  try {
+    // Porosity parameter
+    const double k_p =
+        std::pow(this->porosity_, 3) / std::pow((1. - this->porosity_), 2);
+    // Initialise drag force coefficient
+    VectorDim drag_force_coefficient;
+    drag_force_coefficient.setZero();
+
+    // Check if permeability coefficient is valid
+    for (unsigned i = 0; i < Tdim; ++i) {
+      if (k_p > 0.)
+        drag_force_coefficient(i) =
+            porosity_ * porosity_ * 9.81 *
```

Comment:
Shouldn't the value 9.81 change depends on gravity? Can we pass `gravity_` and then use `gravity_.squaredNorm()`